### DISCUSSION TITLE
不可變資料型態缺點

### DISCUSSION THREAD

**Discussion by sean1325418** [1101-GS4719-C/Python-Community-GS4719-C#383] [Q&A]
今天上課教到python有一種資料型態是不可變的，就是重新設定變數時會把id指向另一個記憶體區域，而原本的區域還是會保存。上課說這樣的優點是不改變記憶體的值可以省CPU時間，可是這樣當程式很大的時候記憶體位置不會不夠嗎。如果不夠是會刪掉舊的值還是程式就直接error了。

**Comment by pekoranoinu** [comment]
記憶體不夠會直接當掉-.-

**Comment by ThomasMeng0515** [comment]
不可變資料型態可以減少重複的值對記憶體空間的佔用，但如果是要用到還未使用過的值的時候，便需要額外使用另一塊記憶體，如此便會降低執行效率，我認為當程式的記憶體不足時，程式便會停止執行。

**Comment by boson13579** [comment]
java 的 string 也是不可變的挖. java 甚至還有 string pool 來存string, 至於記憶體不夠的話....我不知道python 會怎麼處理拉, 不過java 有 garbage collection 機制, 記憶體不夠的時候, 就會回收記憶體空間了, python 應該也有它的方法ㄅ, 直接當掉的話感覺蠻....蠢的XD

**Comment by RogelioKG** [comment]
寫過 C / C++ 的應該都知道，要自己 malloc() & free() / new & delete 真的是一件很麻煩也很瑣碎的事情 🥹，一不小心還會記憶體洩漏 (memory leak)。
(不過 modern C++ 有一些新增特色輔助，例如 smart pointer，物件的去留終於不必須都得由程式設計師費心，而是可以透過這項工具自動管理)

**Comment by RogelioKG** [comment]
Python 也有自己的一套記憶體管理方式：🗑️ **垃圾回收機制** (garbage collection)，可喜可賀的是，它是內建的（歐耶）。
垃圾回收機制會自動管理和回收不再使用的記憶體，防止記憶體洩漏 (memory leak) ，也能更有效使用系統的資源。
釋放物件的判定有兩種方法：
1. **引用計數** (reference counting)
2. **迴圈引用** (reference cycle)

詳細內容可以參考這篇文章：
https://hackmd.io/@MaDi/B1ZD3dtEu

**Comment by RogelioKG** [comment]
當然，我並不是說手動管理記憶體是糟糕的。如果程式設計師能夠清楚地知道自己程式中所有物件的建立和解構時機，手動管理確實可以提升程式碼的效率。但不可避免的，手動管理記憶體會帶來一些潛在的風險和複雜性。

**Comment by RogelioKG** [comment]
然而，現今的軟體開發已走向快節奏且短週期的模式。在這樣的情況下，比起直接使用這些複雜且可能隱含記憶體安全疑慮的功能，我們需要一個更安全的介面來操作這些功能，並在此同時保證記憶體的安全性。這也是 Rust 近年大行其道的原因，因為它主打的就是記憶體安全 (memory safety) 的特色。
